In [1]:
import pandas as pd
import numpy as np
import requests
import json
import tweepy
import os

## Gathering Data

In [2]:
twitarch_df = pd.read_csv('twitter-archive-enhanced-2.csv')

In [3]:
# https://developer.twitter.com/en/docs/authentication/guides/authentication-best-practices
consumer_key = os.environ.get("CONSUMER_KEY")
consumer_secret = os.environ.get("CONSUMER_SECRET")

# from http://docs.tweepy.org/en/latest/getting_started.html
# using Oauth 2
auth = tweepy.AppAuthHandler(consumer_key, consumer_secret)
# https://knowledge.udacity.com/questions/66949#66975
api = tweepy.API(auth_handler = auth,
                 parser = tweepy.parsers.JSONParser(),
                 wait_on_rate_limit = True,
                 wait_on_rate_limit_notify = True)

```
Rate limit reached. Sleeping for: 607
Rate limit reached. Sleeping for: 606
```

### code above ran for 32 minutes
I started execution at 14:30 on 2020-08-24 (Aug 24 2020). It finished around 15:02

In [4]:
# get the info from json object that will be merged with original df
added_features = []

with open('tweet_json.txt', 'r') as file:
    for item in file:
        data = json.loads(item)
        added_features.append({'tweet_id': data['id'],
                               'favorite_count': data['favorite_count'],
                               'retweet_count': data['retweet_count']})

added_fields = pd.DataFrame(added_features)

### The project details gave 2 different URLs for the image predictions data. I will compare the 2 below.

In [5]:
# project gave us 2 different links. They are identical
!diff -s image_predictions.tsv image_predictions_2.tsv

Files image_predictions.tsv and image_predictions_2.tsv are identical


### Since the 2 image_predictions files are identical, I will use the file extracted from first URL.

In [6]:
imagepred_df = pd.read_csv('image_predictions.tsv', sep='\t')

-----------------------

## Assess Data

In [7]:
twitarch_df.info()
twitarch_df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356 

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

In [8]:
# Are all the tweet_ids unique?
print(f"There are {twitarch_df.shape[0]} entries and there are {twitarch_df.tweet_id.nunique()} unique tweet_ids")

There are 2356 entries and there are 2356 unique tweet_ids


All tweet_ids are unique since there are 2356 entries and there are 2356 unique IDs

In [9]:
added_fields.head()

tweet_id  favorite_count  retweet_count
0  892420643555336193           35783           7568
1  892177421306343426           30896           5607
2  891815181378084864           23249           3714
3  891689557279858688           39055           7746
4  891327558926688256           37301           8346

In [10]:
imagepred_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   int64  
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [11]:

imagepred_df.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

In [17]:
twitarch_df[twitarch_df['text'].str.contains('doggos|puppers|puppos|floofers') & (twitarch_df['pupper'] != 'pupper')]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
83    876537666061221889                    NaN                  NaN   
268   841439858740625411                    NaN                  NaN   
274   840698636975636481           8.406983e+17         8.405479e+17   
296   837366284874571778                    NaN                  NaN   
302   836648853927522308                    NaN                  NaN   
475   816062466425819140                    NaN                  NaN   
477   815990720817401858                    NaN                  NaN   
798   772877495989305348                    NaN                  NaN   
934   753420520834629632                    NaN                  NaN   
946   752568224206688256                    NaN                  NaN   
987   749036806121881602                    NaN                  NaN   
993   748575535303884801                    NaN                  NaN   
1027  746056683365994496                    NaN                  NaN   
1093  737310737551491075                    NaN                  NaN   
1120  731156023742988288                    NaN                  NaN   
1220  714606013974974464                    NaN                  NaN   
1228  713900603437621249                    NaN                  NaN   
1254  710658690886586372                    NaN                  NaN   
1265  709901256215666688                    NaN                  NaN   
1351  704054845121142784                    NaN                  NaN   
1516  690959652130045952                    NaN                  NaN   
1634  684225744407494656           6.842229e+17         4.196984e+09   
1635  684222868335505415                    NaN                  NaN   
1636  684200372118904832                    NaN                  NaN   
1643  683857920510050305                    NaN                  NaN   
1710  680583894916304897                    NaN                  NaN   
1712  680494726643068929                    NaN                  NaN   
1743  679405845277462528                    NaN                  NaN   
1826  676440007570247681                    NaN                  NaN   
1843  675853064436391936                    NaN                  NaN   
1847  675820929667219457                    NaN                  NaN   
1862  675432746517426176                    NaN                  NaN   
1900  674664755118911488                    NaN                  NaN   
1928  674045139690631169                    NaN                  NaN   
2141  669993076832759809                    NaN                  NaN   

                      timestamp  \
83    2017-06-18 20:30:39 +0000   
268   2017-03-14 00:04:30 +0000   
274   2017-03-11 22:59:09 +0000   
296   2017-03-02 18:17:34 +0000   
302   2017-02-28 18:46:45 +0000   
475   2017-01-02 23:23:48 +0000   
477   2017-01-02 18:38:42 +0000   
798   2016-09-05 19:22:09 +0000   
934   2016-07-14 02:47:04 +0000   
946   2016-07-11 18:20:21 +0000   
987   2016-07-02 00:27:45 +0000   
993   2016-06-30 17:54:50 +0000   
1027  2016-06-23 19:05:49 +0000   
1093  2016-05-30 15:52:33 +0000   
1120  2016-05-13 16:15:54 +0000   
1220  2016-03-29 00:12:05 +0000   
1228  2016-03-27 01:29:02 +0000   
1254  2016-03-18 02:46:49 +0000   
1265  2016-03-16 00:37:03 +0000   
1351  2016-02-28 21:25:30 +0000   
1516  2016-01-23 18:09:53 +0000   
1634  2016-01-05 04:11:44 +0000   
1635  2016-01-05 04:00:18 +0000   
1636  2016-01-05 02:30:55 +0000   
1643  2016-01-04 03:50:08 +0000   
1710  2015-12-26 03:00:19 +0000   
1712  2015-12-25 21:06:00 +0000   
1743  2015-12-22 20:59:10 +0000   
1826  2015-12-14 16:34:00 +0000   
1843  2015-12-13 01:41:41 +0000   
1847  2015-12-12 23:34:00 +0000   
1862  2015-12-11 21:51:30 +0000   
1900  2015-12-09 18:59:46 +0000   
1928  2015-12-08 01:57:39 +0000   
2141  2015-11-26 21:36:12 +0000   

                                                 source  \
83    <a href="http://twitter.com/download/iphone" r...   
268   <a href="http://twitter.com/download/i

In [18]:
twitarch_df['text'].str.contains('doggo|pupper|puppo|floofer').sum()

399

In [21]:
twitarch_df[(twitarch_df['text'].str.contains('^RT', regex=True)) & (twitarch_df['retweeted_status_id'].isnull())]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
1286  708400866336894977                    NaN                  NaN   
1860  675489971617296384                    NaN                  NaN   

                      timestamp  \
1286  2016-03-11 21:15:02 +0000   
1860  2015-12-12 01:38:53 +0000   

                                                 source  \
1286  <a href="http://vine.co" rel="nofollow">Vine -...   
1860  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
1286  RT if you are as ready for summer as this pup ...                  NaN   
1860  RT until we find this dog. Clearly a cool dog ...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
1286                       NaN                        NaN   
1860                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
1286                      https://vine.co/v/iHFqnjKVbIQ                12   
1860  https://twitter.com/dog_rates/status/675489971...                10   

      rating_denominator  name doggo floofer pupper puppo  
1286                  10  None  None    None   None  None  
1860                  10  None  None    None   None  None